examples/vae_text/vae_train.py

```diff
@@ -0,0 +1,477 @@
+
+        nll = rc_loss + kl_weight * kl_loss
+
+        ret = {
+            "nll": nll,
+            "kl_loss": kl_loss,
+            "rc_loss": rc_loss,
+            "lengths": seq_lengths
+        }
+
+        return ret
+
+    def _embed_fn_rnn(self, tokens: torch.LongTensor) -> Tensor:
+        r"""Generates word embeddings
+        """
+        embedding = self.decoder_w_embedder(
+            tokens)
+        latent_z = self._latent_z
+        if len(embedding.size()) > 2:
+            latent_z = torch.unsqueeze(latent_z, 0)
+            latent_z = latent_z.repeat([tokens.size(0), 1, 1])
+        return torch.cat([embedding, latent_z], dim=-1)
+
+    def _embed_fn_transformer(self,
+                              tokens: torch.LongTensor,
+                              positions: torch.LongTensor) -> Tensor:
+        r"""Generates word embeddings combined with positional embeddings
+        """
+        output_p_embed = self.decoder_p_embedder(positions)
+        output_w_embed = self.decoder_w_embedder(tokens)
+        output_w_embed = output_w_embed * \
+            self._config.hidden_size ** 0.5
+        output_embed = output_w_embed + output_p_embed
+        return output_embed
+
+    def decode(self,
+               helper: Optional[Helper],
+               text_ids: Optional[torch.LongTensor] = None,
+               latent_z: Optional[Tensor] = None,
+               seq_lengths: Optional[Tensor] = None,
+               max_decoding_length: Optional[int] = None) \
+            -> Union[BasicRNNDecoderOutput, TransformerDecoderOutput]:
+
+        self._latent_z = latent_z
+        fc_output = self.mlp_linear_layer(latent_z)
+
+        if self._config.decoder_type == "lstm":
+            decoder_initial_state_size = (self.decoder.cell.hidden_size,) * 2
+            decoder_states = torch.split(
+                fc_output, decoder_initial_state_size, dim=1)
+            outputs, _, _ = self.decoder(
+                initial_state=decoder_states,
+                inputs=text_ids,
+                helper=helper,
+                sequence_length=seq_lengths,
+                max_decoding_length=max_decoding_length)
+        else:
+            decoder_initial_state_size = torch.Size(
+                [1, self._config.dec_emb_hparams["dim"]])
+            decoder_states = torch.reshape(
+                fc_output, (-1, ) + decoder_initial_state_size)
+            outputs = self.decoder(
+                inputs=text_ids,
+                memory=decoder_states,
+                memory_sequence_length=torch.ones(decoder_states.size(0)),
+                helper=helper,
+                max_decoding_length=max_decoding_length)
+        return outputs
+
+
+class VAEData(tx.data.MonoTextData):
+    r""" Wrap `tx.data.MonoTextData` to implement data preprocessing.
+    """
+    def process(self, raw_example: List[str]) -> List[str]:
+        # Truncates sentences and appends BOS/EOS tokens.
+        words = super().process(raw_example[1:-1])
```

Comment:
I think the extra blank token problem is already fixed in the master branch? Is this still necessary?

Reply:
Yes, it is redundant now.